f-form@v0.3.0 - add field validations

FayeCarter opened the pull request:
### Added
- Add field validations.
![image](https://user-images.githubusercontent.com/24740015/146033093-8df94e7b-ed2a-4f75-af6c-4a8cb374f6b9.png)

FayeCarter committed:
generate component

FayeCarter committed:
Add basic form components

FayeCarter committed:
Add validations

FayeCarter committed:
Update readme

matthewhardern committed:
approval process for circle ci (#1550)

dandel10n committed:
f-header@9.1.0 - jet logo (#1551)

* f-header@9.1.0 - jet logo

* f-header@8.1.0 - fix arrow icon size

* f-header@8.1.0 - renamed the prop

* f-header@8.1.0 - unit test for the new prop

FayeCarter committed:
Update tests

FayeCarter committed:
update version

xander-marjoram commented on packages/components/organisms/f-form/README.md:
```diff
@@ -96,7 +110,8 @@ formData: {
 | ----- | ----- | ------- | ----------- |
 | `formData` | Object | - | Object containing button text and form field data. Should be structured as above. |
-| `isFormSubmitting` | Boolean | - | Allows the parent component to set a loading spinner on the submit button while any asynchronous calls are carried out by the parent |
+| `locale` | String | `en-gb` | Sets the translation file to use. |
```
Nit
```suggestion
| `locale` | String | `en-GB` | Sets the translation file to use. |
```

xander-marjoram commented on packages/components/organisms/f-form/src/components/_tests/Form.test.js:
```diff
@@ -43,21 +18,66 @@ describe('Form', () => {
+        const forms = wrapper.findAll('form');
+
+        // Assert
+        expect(forms.length).toBe(1);
```
It it likely that there will be two or more forms? Not sure it's worth checking for that

xander-marjoram commented on packages/components/organisms/f-form/src/components/Form.vue:
```diff
@@ -67,33 +105,112 @@ export default {
+
+            return invalidFieldCount === 1
+                ? this.copy.errorMessages.singleFieldError
+                : this.copy.errorMessages.multipleFieldErrors.start + invalidFieldCount + this.copy.errorMessages.multipleFieldErrors.end;
```
Instead of enforcing this kind of sentence structure could you try something like this, where you have a template and the number gets passed in?
https://kazupon.github.io/vue-i18n/guide/formatting.html#support-ruby-on-rails-i18n-format

xander-marjoram commented on packages/components/organisms/f-form/src/components/_tests/Form.test.js:
```diff
@@ -76,17 +96,205 @@ describe('Form', () => {
+                jest.clearAllMocks();
+            });
+
+            it('should call `.touch()` ', () => {
```
Do we need to specifically know that it has called `touch`, or is it better to check that the fields are now dirty?

xander-marjoram replied:
Having thought about this some more, I'm not sure how easy this will be whilst mocking the `$v` object without just testing your mock. I think what you've got now is fine

xander-marjoram commented on packages/components/organisms/f-form/src/components/_tests/Form.test.js:
```diff
@@ -76,17 +96,205 @@ describe('Form', () => {
+                $v.$invalid = true;
+
+                // Act
+                const wrapper = shallowMount(FForm, { propsData, mocks: { $v } });
```
Could setting `$invalid` like this potentially carry state across tests? I wonder if something like this might be a bit more robust:

(Feel free to trim it down a bit visually if you think it takes up too many lines)

```suggestion
                // Arrange & Act
                const wrapper = shallowMount(FForm, {
                    propsData,
                    mocks: {
                        $v: {
                            ...$v,
                            $invalid: true
                        }
                    }
                });
```

xander-marjoram commented on packages/components/organisms/f-form/src/components/_tests/Form.test.js:
```diff
@@ -76,17 +96,205 @@ describe('Form', () => {
+            let wrapper;
+
+            beforeEach(() => {
+                isFormValidSpy = jest.spyOn(FForm.methods, 'isFormValid');
```
Instead of spying on the method, I think it would be a little cleaner to mock the underlying value of `$v.$invalid`.

Spies are perfect when you want to test that something is being called, but also maintain the original functionality.
As you're changing the functionality as part of the test (by mocking the return value), it's probably not an ideal usage.

xander-marjoram commented on packages/components/organisms/f-form/src/components/_tests/Form.test.js:
```diff
@@ -76,17 +96,205 @@ describe('Form', () => {
+                wrapper.vm.isValidPhoneNumber();
+
+                // Assert
+                expect(isValidPhoneNumberSpy).toHaveBeenCalledWith(formData.formFields.mobileNumber.value, 'en-GB');
```
For example, this is a good use of a spy.
It's an external method so you can't just stub it with `jest.fn` (unless you use a module mock which can get a bit hairy), and you're also using an expectation like `toHaveBeenCalledWith`.

xander-marjoram commented on packages/components/organisms/f-form/src/components/_tests/Form.test.js:
```diff
@@ -76,17 +96,205 @@ describe('Form', () => {
+        });
+
+        describe('hasInvalidErrorMessage ::', () => {
+            it('should return true if passed field has `invalidValidationMessages`', () => {
```
Is `invalidValidationMessages` a variable somewhere?

xander-marjoram commented on packages/components/organisms/f-form/src/components/_tests/FormField.test.js:
```diff
@@ -0,0 +1,459 @@
+        describe('updateField ::', () => {
+            it('should emit `updated` event with fieldName and field value', () => {
+                // Arrange
+                const wrapper = shallowMount(FormField, { propsData, provide: { $v } });
```
I think in all other places, provide is a function, does this matter? Also could be worth pulling the whole thing out into a variable to save a few lines, etc.

xander-marjoram commented on packages/components/organisms/f-form/src/components/_tests/FormField.test.js:
```diff
@@ -0,0 +1,459 @@
+        });
+
+        describe('ariaLabel ::', () => {
+            it('should return the mobile number with spaces after every character', () => {
```
Do we need to handle any bad values?

xander-marjoram commented on packages/components/organisms/f-form/src/components/_tests/FormField.test.js:
```diff
@@ -0,0 +1,459 @@
+        });
+
+        describe('formFieldBlur ::', () => {
+            it('should call `touch` for validation', () => {
```
Again, it's probably worth testing the result (form field is now dirty) rather than the implementation (`touch` method has been called).

xander-marjoram replied:
See other comment on other thread

xander-marjoram commented on packages/components/organisms/f-form/src/tenants/en-GB.js:
```diff
@@ -1,4 +1,10 @@
     locale: 'en-GB',
-    text: 'I am a VForm Component (GB)'
+    errorMessages: {
+        singleFieldError: 'There is 1 error in the form',
```
Maybe worth looking into pluralisation too: https://kazupon.github.io/vue-i18n/guide/pluralization.html#pluralization

xander-marjoram committed:
f-contact-preferences@0.11.0 - Add info and error logging (#1555)

* f-contact-preferences@0.11.0 - Add info and error logging

* f-contact-preferences@0.11.0 - Update unit tests

* Minor: Add comment

* Minor: Reinstate storeOverride in tests

* Minor: Check for tags

Co-authored-by: Xander Marjoram <xander.marjoram@justeattakeaway.com>

siggerzz committed:
fozzie-components@v5.7.0 - Modify Storybook config to prevent out of memory exception when running global builds. (#1561)

* Serve storybook instead of build

* Update Storybook dep to fix memory leak

* Fix storybook memory leak and revert config

* Test optimisation changes

* Add changelogs

* Remove approval process for jobs

LTurns committed:
#trivial - small fixes for pages component tests (#1562)

* trivial fixes to pages tests

* reverted file name

* file route change

* rebased with master

Co-authored-by: Lizzie Turney <lizzie.turney@just-eat.com>

LTurns committed:
* fixes for atom components

* added isClickable tests

* changed function name for consistency

* trivial fixes to pages tests

* reverted file name

* changed filename to image-tile

* fixed requested changes

* fixed file name

* file route change

* rebased with master

* corrected button name

Co-authored-by: Lizzie Turney <lizzie.turney@just-eat.com>

AndyMayJE committed:
f-account-info@0.13.0 - Add error handling and card (#1557)

- Error handling and error card using `f-card-with-content`
- `info` and `error` logs to preferences fetch and save requests

LTurns committed:
* small fixes to molecules component tests

* small fixes for cookie-banner

* updated content cards

* refactored organism component tests

* trivial fixes to pages tests

* reverted file name

* file route change

* removed unneeded selector file

* updated name in tests

* resolved changes in package.json

Co-authored-by: Lizzie Turney <lizzie.turney@just-eat.com>

Commit:
f-account-info@0.14.0 - Adding analytics. (#1568)

* f-account-info-analytics - Basic commit.

* f-account-info-analytics - Implementation and some basic tests.

* f-account-info-analytics - More tests.

* f-account-info-analytics - Updating changelog and package.json

* f-account-info-analytics - Added more tests.

* f-account-info-analytics - Fixing tests after merge.

* f-account-info-analytics - Amending tests.

* f-account-info-analytics - Reverting changes to fields since validations have now been fixed.

* f-account-info-analytics - Code review improvements.

* f-account-info-analytics - Code review suggestions.

Co-authored-by: Sacha Zvetelman <sacha.zvetelman@justeattakeaway.com>

Commit:
f-checkout@3.13.0 - Split notes v2 (#1506)

* WIP visual tests

* Visual test

* Add version bump

* Add age verification import

* Update CHANGELOG.md

* Update package.json

* Remove commented out code

* Remove old initialisation

* Split Notes v2

* Fix help text on notes

* Update package version

* Remove console.logs

* Remove split notes browser test for now

* Split notes with feature toggling

* Bump version

* Update stories file to pass in checkout features

* Fix broken component test

* Rename test

* Code review changes

* Add split notes component tests

* Update order placement keys in test

* Update noteForDelivery key

* Fix browser test for order

* Update order placement keys

* Restore precommit file

* Correct shape of mapNotesFromApi function

* Clean up orderplacement keys

* Revert pre-commit change

* Add back plastic cutlery placeholder

* Code review changes pt 1

* remove console.log

* Update packages/components/pages/f-checkout/src/components/_tests/Accordion.test.js

Co-authored-by: Xander Marjoram <xander.marjoram@just-eat.com>

* Update date in changelog

* Use getters instead of computed

* removed formatted space from visual test

* use it.each for accordion toggle test

* Code review changes

* Add service type back to mapState

* Added back label description

Co-authored-by: Xander Marjoram <xander.marjoram@just-eat.com>

FayeCarter committed:
Update changelog

FayeCarter committed:
Fix test

AlixRichards committed:
f-image-tile@0.2.0  - Add toggle functionality (#1566)

* Add initial component markup
Add props to storybook story

* Added displayText prop to story
Updated template to use displayText prop

Added $emit when filter is toggled
Added function to create emit object

WIP updating isSelected prop

* Added tests for methods, props and wip watcher test

* Added npm package version change
Added unique id for input
Added image as prop
WIP watch isSelected()
Commented out watch test

* Added test ids
Added props tests
Added watch test

* Updated stories to use controls instead of knobs
Added a demo story to show imageTile within a container which tests the $emit event

* removed unneeded prop from test container
Added changelog

* removed unneeded default value
Updated story name

* Updated storybook title
Updated component test file with data-test-id

* renamed data-test-id to image-tile-component
Removed unused prop in storybook

* Moving end of tag on to same line as last attribute

* Corrected spacing within css
Updated aria-hidden label conditions

* Update packages/components/atoms/f-image-tile/src/components/ImageTile.vue

Co-authored-by: Anastasiya <a.tyshkavets@justeattakeaway.com>

* Update packages/components/atoms/f-image-tile/src/components/ImageTile.vue

Co-authored-by: Anastasiya <a.tyshkavets@justeattakeaway.com>

* Update packages/components/atoms/f-image-tile/src/components/ImageTile.vue

Co-authored-by: Anastasiya <a.tyshkavets@justeattakeaway.com>

* Update packages/components/atoms/f-image-tile/src/components/ImageTile.vue

Co-authored-by: Anastasiya <a.tyshkavets@justeattakeaway.com>

* Updated tests to use it.each()

* renamed filterEmitObject to createEmitObject

* Renamed isFilterSelected to isToggleSelected

Co-authored-by: Alix Richards <alix.richards@justeattakeaway.com>
Co-authored-by: Anastasiya <a.tyshkavets@justeattakeaway.com>

dandel10n committed:
f-header@9.2.0 - remove ml theme style overrides (#1570)

* f-header@9.2.0 - remove ml theme override

* fozzie-components@5.8.0 - fozzie package update

* f-header@9.2.0 - remove NO and DK copy

* f-header@9.2.0 - fix logo tests

FayeCarter committed:
Update version

FayeCarter committed:
Update version

FayeCarter committed:
Fix tests

FayeCarter committed:
f-form@0.1.0 - generate component (#1553)

* generate component

* Bump version

* Update changelog

* Fix test

* Fix test

* Fix tests

Co-authored-by: Faye Carter <faye.carter@justeattakeaway.com>

FayeCarter committed:
Merge branch 'master' of https://github.com/justeat/fozzie-components

FayeCarter committed:
Update readme

FayeCarter committed:
Use array

FayeCarter committed:
Updates from review

matthewhardern committed:
Changes to Circle CI config - #globalconfig  (#1565)

* changes to test circle ci

* circle ci changes 2

* restore cache add back

* Persist atoms to workspace

* Run job multiple times

* Add linting scripts

* fix f-globalisation package.json

* Fix linting for services

* Depend on tools build for molecules

* Persist node modules to workspace

* Cache yarn cache

* Update workflow / cache

* Revert to working persist config

* Revert to working config with persited dist

* single build job with matrix unit tests

* Test job for molecules

* Add new jobs for organisms/pages/templates

* Add test scripts for atoms/molecules/organisms

* Fix browser test build name

* Fix browser build

* Fix script names

* re-enable build steps

* Move Danger step

* Add pages scripts

* Add bundlewatch check

* Add bundlewatch check

* Restore dist instead of build for deploy-storybook job

* Save cache of @justeat namespaced packages

* Fix package.json scripts

* PR comments

* Remove old ci:lint scripts

* Add lint script

* Add lint script

Co-authored-by: Ben Siggery <bensiggery28@gmail.com>

FayeCarter committed:
Merge branch 'master' of https://github.com/justeat/fozzie-components

FayeCarter committed:
Merge branch 'master' of https://github.com/justeat/fozzie-components

FayeCarter committed:
Merge branch 'master' of https://github.com/justeat/fozzie-components

FayeCarter committed:
f-form@0.2.0 - Add basic components (#1554)

* generate component

* Add basic form components

* Update readme

* Update changelog

* Fix test

* Update version

* Update readme

* Fix tests

* Fix field

* Fix typo

* Improve test

* Readme and dependency update

* Add form data validations

* Add constants

* Adding peer dependencies external config to vue config

* Adding npm scripts for organism tests

* Package.json and changelog for global changes

* remove visual tests

* Update packages/components/organisms/f-form/README.md

Co-authored-by: Ashley Watson-Nolan <ashley.watson-nolan@justeattakeaway.com>

Co-authored-by: Faye Carter <faye.carter@justeattakeaway.com>
Co-authored-by: ashleynolan <nolly00@gmail.com>
Co-authored-by: Ashley Watson-Nolan <ashley.watson-nolan@justeattakeaway.com>

FayeCarter committed:
Merge branch 'master' of https://github.com/justeat/fozzie-components

FayeCarter committed:
Remove extra globalisation

zlatinivanov-je commented on packages/components/organisms/f-form/README.md:
```diff
@@ -73,17 +73,17 @@ export default {
-        * **label** - the field label
-        * **validationMessages** - an object that can containin two types of validation messages:
+        * **label** - the field's displayed text label
+        * **validationMessages** - an object that can contain two types of validation messages:
```
This is interesting, however it seems to be limiting to `required` and `invalid` without relation to the validation rule(s). We build something similar recently and had to build a field validity object, also observe validation order... 
It can easily get complicated. Happy to chat to you about that if that is the on the cards here.

FayeCarter replied:
Yeah would be good to see what you've already built to see if there's anything that could be improved. 
At the moment it works out the validation message here https://github.com/justeat/fozzie-components/pull/1556/files#diff-fdc8192e0a3e68bb066e4093bc5964e0edec3662062fca586688c40fc5a9ca34R121 wheich is the same as is currently used in checkout

FayeCarter committed:
Updates to readme

FayeCarter committed:
Revert rebase issue

FayeCarter committed:
Revert rebase issue

FayeCarter committed:
Clean up tests

Remove extra globalisation

Revert rebase issue

Revert rebase issue

FayeCarter committed:
Merge branch 'f-form-validation' of https://github.com/justeat/fozzie-components into f-form-validation

FayeCarter committed:
Increase Bundlewatch size
